### PR TITLE
DB-10049 user-defined statement separator in sqlshell.

### PR DESCRIPTION
**Pull request by hatyo**

### CODE REVIEW

**Review by martinrupp — APPROVED** [review]

**Comment by martinrupp** [comment]
verified we can do now things like
```
SET TERMINATOR '%';

CREATE TRIGGER mytrig2
AFTER UPDATE OF a,b
ON t1
REFERENCING OLD AS OLD NEW AS NEW
FOR EACH ROW
WHEN (EXISTS (SELECT 1 from t2 where t2.a = OLD.a))
BEGIN ATOMIC
    SELECT * FROM SYS.SYSTABLES;
    SIGNAL SQLSTATE '87101' SET MESSAGE_TEXT = 'a ';
END%
```

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java [hunk 1]
```diff
@@ -692,206 +667,185 @@ private void checkScrollableCursor(ResultSet rs, String operation)
...
+    }
+
+    public void setOmitHeader(boolean omitHeader) {
+        this.omitHeader = omitHeader;
+    }
+
+    public void setTerminator(String terminator) {
+        if (terminator.length() != 1 || !StatementFinder.isValidTerminator(terminator.charAt(0))) {
+            out.println(langUtil.getTextMessage("IJ_TerminatorInvalid", terminator));
+        }
```
Comment (martinrupp) [inline db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:847]: `return` missing

**Review by martinrupp — CHANGES_REQUESTED** [review]

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java (hunk 1 above)
Comment (martinrupp) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:847]: e.g. this fails

```
set terminator ',';
show tables;
```

**Review by arnaud-splice — CHANGES_REQUESTED** [review]
The following scenario doesn't seem to work:
- `set terminator 'a'; create schema foo a` -> Create schema does not see 'a' as a terminating character

Additionally
- `set terminator '~~'` raises an error but still silently sets the terminator to the first character of `~~`
We should either say that in the error message or not set it at all.

**Review by dgomezferro — COMMENTED** [review]
Looks good to me except for the already raised issues.

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java (hunk 1 above)
Comment (hatyo) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/utilMain.java:847]: Thanks Martin, fixed.

**Comment by hatyo** [comment]
@arnaud-splice:

> set terminator 'a'; create schema foo a -> Create schema does not see 'a' as a terminating character

Thanks! solving this problem (and other related problems) requires changing the way we pre-parse incoming stream of characters before passing it to respective parsers, for more information please have a look at my comment on respective JIRA items.

I pushed a new commit that disallows users from choosing an alphabet character as a statement separator since I don't think it is a common use case anyway.

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/StatementFinder.java [hunk 2]
```diff
@@ -31,549 +31,567 @@
...
+        assert isValidTerminator(terminator);
+        this.terminator = terminator;
+    }
+
+    private static boolean isAsciiCharacter(char c) {
+        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+    }
+
+    public static boolean isValidTerminator(char terminator) {
+        return !isAsciiCharacter(terminator) && !DISALLOWED_TERMINATORS.contains(terminator);
```
Comment (martinrupp) [inline db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/StatementFinder.java:591]: (minor comment) also ; is an ASCII character... i think you mean Character.isLetter(terminator)

**Review by martinrupp — APPROVED** [review]

File: db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/StatementFinder.java (hunk 2 above)
Comment (arnaud-splice) [reply db-tools-ij/src/main/java/com/splicemachine/db/impl/tools/ij/StatementFinder.java:591]: +1

**Review by arnaud-splice — APPROVED** [review]
Good change! Please address Martin's comment and merge the latest master to resolve the build issue

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by martinrupp** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**